### ISSUE TITLE
Why `findValues` is a `Seq` and not a `Set`?

### ISSUE THREAD

**Issue by joan38** [lloydmeta/enumeratum#55]
All in the title

**Comment by lloydmeta** [comment]
Hi there,

Thanks for your interest in this project. To answer your question:

1. We want to preserve order so we can find the index of each member, and `Set` cannot be relied on to preserve order.
2. If we are worried about duplicates, the Scala compiler already takes care of that for us because you cannot define 2 objects with the same name in the same scope.
3. I wanted to have constant-time by-index access to `value` members, hence `findValues` actually will return an `IndexedSeq`

I hope that answers your question.

**Comment by joan38** [comment]
Thanks @lloydmeta for your answer. That makes sense.

**Comment by randomstatistic** [comment]
Regarding the second point, (which is why I found this issue) that's only true if your entryName is the object name, yes? If that's not true, it looks completely possible to define duplicate "enums". There are no errors, and only one (at "random") of the objects is findable using withName().
There's a uniqueness check in the ValueEnums, but not for "string-like" enums.

**Comment by lloydmeta** [comment]
@randomstatistic Hi there, and thanks for your interest in this project.  The difference is subtle, but I think that isn't a case of duplicate enums (same identifier in code), but rather on of duplicate enum names.  

A bit of background is probably in order. One of the goals of Enumeratum is to provide an enum implementation that is similar to the standard lib's `Enumeration`, but better with regards to type-safety and flexibility. This is so that it could be (as much as possible) as straight forward swap between the two.  Note that you could do something like this with the built-in `Enumeration`

```scala
object Numbers extends Enumeration {

  val threeString = "Three"
  val One = Value("One")
  val Two = Value("One") // same issue we have in Enumeratum
  val Three = Value(threeString) // Non-literal names allowed!
}

scala> Numbers.withName("One")
res1: Numbers.Value = One

scala> Numbers.withName("One").id
res2: Int = 0
```

I later introduced `ValueEnum` to this lib as my own take on what value-based enums should be like.  Since it doesn't exist in the standard lib, I was free to model it as I saw fit: for example, adding a uniqueness-check on the values of the enums, which comes a restraint: the values must be declared as constant literals.

So, in short, I think the `Enum` of Enumeratum will likely not add such a check because it needs to stay flexible and similar enough to that of `Enumeration`.  However, it is entirely possible (and quite-easy code-wise) to add a `StringValueEnum` that has the same uniqueness checks as the number-based `ValueEnums` :)

**Comment by randomstatistic** [comment]
@lloydmeta Thanks for the detailed response, I was very pleased to find what you've built here.

`ValueEnum` looks sealed, so it'd take a PR to add another value type. (which I'd be happy to do, although perhaps not this week) 

I do very much prefer to think of an enum being equivalent to a constant. If it represents any "value" at all, then those values should match the same uniqueness constraints as the enums themselves. The enum IS the value.

There might be some ambiguity around the distinction between a "value" and a "name" though. An `Enumeration`'s value is an Int. It also has an optional name. An `EnumEntry`'s value is a String. By default it's the object name, but you're encouraged to override it such that the value becomes a name **too**. (Yes, there's an index in there as well, but the `entryName` is not optional, so I don't consider that quite the same)
I think this is at the root of why I was surprised that there was no uniqueness check. I didn't want the enum-object's name to be tied to the enum value, so I just set the entryName to what I wanted the value to be, expecting that if there was uniqueness before, there must be still.

That said, I think the `Enumeration` behaviour should be considered a bug too - either it shouldn't allow names at all, or it shouldn't allow non-distinct names any more than you can have non-distinct values. The function named `withName()` certainly implies some assurance that you can get a given enum value back again.

Anyway, I certainly prefer your interpretation to that one, and I think I would be perfectly satisfied with a `StringValueEnum` approach. I can understand wanting to be bug-compatible. :)

**Comment by randomstatistic** [comment]
Oh, and apologies to @joan38 for hijacking his thread.

**Comment by joan38** [comment]
No worries mate ;)
